### PR TITLE
Incremental refresh for refresh_continuous_aggregate()

### PR DESCRIPTION
**Pull request by melihmutlu**
refresh_continuous_aggregate() can now refresh incrementally in batches,
matching the behavior the continuous aggregate policy already supports.
Forced refresh is also incremental by default.

Options available for incremental refresh with policies can be set for
manual refresh via JSONB options:

  - `buckets_per_batch`         (default 10, as policies)
  - `max_batches_per_execution` (default 0 = no limit)
  - `refresh_newest_first`

Usage:

  Without options, it behaves similar to a refresh policy:

      CALL refresh_continuous_aggregate('cagg', t1, t2);

### CODE REVIEW

File: tsl/src/continuous_aggs/refresh.c [hunk 1]
```diff
@@ -651,6 +651,7 @@ continuous_agg_refresh(PG_FUNCTION_ARGS)
 	{
 		/* get min time for a cagg depending of the primary partition type */
 		refresh_window.start = cagg_get_time_min(cagg);
+		refresh_window.start_isnull = true;
```
Comment (pnthao) [inline tsl/src/continuous_aggs/refresh.c:771]: Was it a bug that we didn't set start_isnull = true before?

Comment (kpan2034) [reply tsl/src/continuous_aggs/refresh.c:771]: We directly passed it in with `PG_ARGISNULL(1)` in the call to `continuous_agg_refresh_internal`

https://github.com/timescale/timescaledb/pull/9903/changes#diff-cdbea219439b7594e17f13199b3e53fead5761b7ad4c9eb4a58ec3fd54225ac5L672

File: tsl/src/continuous_aggs/refresh.c [hunk 2]
```diff
@@ -624,6 +623,113 @@ continuous_agg_refresh_with_window(const ContinuousAgg *cagg,
...
+		 * In non-batched mode (single batch) the one batch is always the boundary. */
+		bool apply_extend =
+			extend_last_bucket &&
+			(context.refresh_newest_first ? processing_batch == 1 :
+											processing_batch == context.number_of_batches);
+
+		any_refreshed |= continuous_agg_refresh_internal(cagg,
+														 batch_window,
+														 context,
+														 batch_window->start_isnull,
```
Comment (pnthao) [inline tsl/src/continuous_aggs/refresh.c]: Should we also set start_isnull and end_isnull in batch_window here, instead of passing them as separate args?

File: tsl/test/expected/cagg_invalidation.out [hunk 3]
```diff
@@ -1687,6 +1687,8 @@ SET timezone = 'UTC';
 --Do the first refresh and check materialization invalidation log
 call refresh_continuous_aggregate ('test_cagg','2023-12-29 15:00:00', '2026-01-28 15:00:00');
 call refresh_continuous_aggregate ('test_cagg_1d_offset','2023-12-29 18:00:00', '2024-01-01 18:00:00');
+-- Batches of [2024-01-01 ─ 2024-11-01) and [2024-11-01 ─ 2025-09-01) should already been processed by the above refresh
```
Comment (pnthao) [inline tsl/test/expected/cagg_invalidation.out]: I'd suggest adding a bit more info to the comment so it's clearer why the refresh stops at 2005-09-01 instead of going all the way to 2026-01-01:
  -- test_cagg uses 1-month buckets with buckets_per_batch=10. The inscribed window
  -- [2024-01-01, 2026-01-01) spans 24 months, producing two batches:
  --   [2024-01-01, 2024-11-01) and [2024-11-01, 2025-09-01).
  -- Data only exists through Dec 2024, so the split function finds no chunks past
  -- Dec 2024 and stops at Sep 2025 (end of batch 2). The range [2025-09-01, 2026-01-01)
  -- has no data and is left unprocessed, so the upper residual in the mat_inval_log
  -- is 2025-09-01 rather than 2026-01-01 as it would be with single-pass refresh.

File: tsl/test/expected/cagg_invalidation.out [hunk 4]
```diff
@@ -1800,7 +1803,7 @@ WHERE hypertable_id IN (
   SELECT raw_hypertable_id FROM _timescaledb_catalog.continuous_agg WHERE user_view_name = 'test_cagg_1d_offset');
  invalidation_threshold 
 ------------------------
- 2026-01-05 18:00:00+00
```
Comment (pnthao) [inline tsl/test/expected/cagg_invalidation.out:1803]: The reason for the change is a bit more subtle, around the logic that gets and sets the new invalidation threshold
Before the refresh, the existing threshold is 2026-01-01 00:00 (set by a refresh on another cagg with no offset)
Batch 1 ([Jan 6 18:00, infinity]). Because window end is infinity, it is capped at next_bucket_start(max data) =  = Jan 5 18:00. But window start (Jan 6) > capped_end (Jan 5) → no work done.
Batch 2 ([Dec 27 2025 18:00, Jan 6 2026 18:00]). The window has finite end so it doesn't get capped, so the window end of Jan 6 2026 18:00 then replaces the threshold set by the previous batch.

For this test, I think we should use bucket_per_batch = 0, because the intention of the test is to test the capping in case of window end  = null.

File: tsl/test/expected/cagg_invalidation.out [hunk 5]
```diff
@@ -1899,7 +1928,7 @@ ORDER BY 1,2;
           low           |   high    
 ------------------------+-----------
  -infinity              | -infinity
- 2020-01-01 04:00:00+00 | infinity
+ 2020-01-01 16:00:00+00 | infinity
```
Comment (pnthao) [inline tsl/test/expected/cagg_invalidation.out]: This change is similar to the threshold change above, due to the fact that the second batch has a finite window end and moves the threshold forward. I think we can also use bucket_per_batch = 0 for this test, to preserve initial test intention (that the refresh window is cap at bucket boundary of the 4hrs cagg, rather than using that of the 6hrs cagg)

File: tsl/test/expected/cagg_query-18.out [hunk 6]
```diff
@@ -96,11 +96,11 @@ select * from mat_m1 order by sumh, sumt, minl, timec ;
    Sort Key: _materialized_hypertable_2.sumh, _materialized_hypertable_2.sumt, _materialized_hypertable_2.minl, _materialized_hypertable_2.timec
    ->  Append
          ->  Append
-               ->  Seq Scan on _timescaledb_internal._hyper_2_3_chunk
```
Comment (pnthao) [inline tsl/test/expected/cagg_query-18.out:99]: I think we can disable incremental refresh in this test file (cagg_query). I verified the diffs are expected (just chunk name swapping because we refresh newest batch first), but I think incremental refresh could make it harder to observe what the test is intended for.

File: tsl/test/expected/cagg_invalidation_variable_bucket.out [hunk 7]
```diff
@@ -309,9 +309,9 @@ CALL refresh_continuous_aggregate('cagg_dst_daily', '2025-09-01 00:00:00', '2026
 SELECT * FROM cagg_inval_log WHERE cagg_name = 'cagg_dst_daily';
    cagg_name    |      inval_start       |           inval_end           
 ----------------+------------------------+-------------------------------
```
Comment (pnthao) [inline tsl/test/expected/cagg_invalidation_variable_bucket.out:312]: These changes are all because incremental refresh only starts a batch where there is a chunk (and invalidations), while one-pass refresh only looks at invalidations. The change is irrelevant to the purpose of the test, so we can either leave it as is or change to 1 pass refresh for simplicity.

File: tsl/test/expected/telemetry_stats.out [hunk 8]
```diff
@@ -303,25 +303,25 @@ SELECT jsonb_pretty(rels) AS relations FROM relations;
          "num_reltuples": 697                             +
      },                                                   +
      "hypertables": {                                     +
-         "heap_size": 335872,                             +
```
Comment (pnthao) [inline tsl/test/expected/telemetry_stats.out:306]: Incremental refresh newest first changes the order the chunks are created and hence their ids. The test gets the chunk list order by their ids and picks the first one to compress. That explains the output changes. For simplicity, we can use one pass refresh for these test.

File: tsl/test/isolation/specs/cagg_refresh_cleanup_register.spec [hunk 9]
```diff
@@ -220,11 +220,11 @@ step "P1_run_policy" {
 
 # Two refreshes wait for registration, one waits for cleanup before exiting. All blocked on an AccessExclusiveLock on continuous_aggs_jobs_refresh_ranges.
 # None of those refreshes overlaps, so all should succeed.
-permutation "WP_mat_enable" "R2_refresh" "L1_lock" "WP_mat_disable" "R3_refresh" "R4_refresh" "check_locks" "check_jobs" "L1_unlock" "check_locks" "check_jobs"
+permutation "WP_mat_enable" "R2_refresh" "L1_lock" "WP_mat_disable" "R3_refresh"("R2_refresh") "R4_refresh"("R3_refresh") "check_locks" "check_jobs" "L1_unlock" "check_locks" "check_jobs"
```
Comment (kpan2034) [inline tsl/test/isolation/specs/cagg_refresh_cleanup_register.spec]: Looks like there's still a race condition that can cause [tests to fail](https://github.com/timescale/timescaledb/actions/runs/27037678643/job/79805732253?pr=9903)


```suggestion
permutation "WP_mat_enable" "R2_refresh" "L1_lock" "WP_mat_disable" "R3_refresh"("R2_refresh") "R4_refresh"("R3_refresh") "check_locks" "check_jobs" "L1_unlock" "check_locks"("R4_refresh") "check_jobs"
```

**Review by kpan2034 — APPROVED** [review]
One test needs a small fix, other than that LGTM

File: tsl/test/expected/cagg_invalidation_variable_bucket.out (hunk 7 above)
Comment (melihmutlu) [reply tsl/test/expected/cagg_invalidation_variable_bucket.out:312]: Changed it to a single pass refresh

File: tsl/test/expected/cagg_invalidation.out (hunk 4 above)
Comment (melihmutlu) [reply tsl/test/expected/cagg_invalidation.out:1803]: Disabled incremental refresh.

File: tsl/test/expected/cagg_invalidation.out (hunk 5 above)
Comment (melihmutlu) [reply tsl/test/expected/cagg_invalidation.out]: Disabled incremental refresh.

**Review by pnthao — APPROVED** [review]

File: tsl/src/continuous_aggs/refresh.c [hunk 10]
```diff
@@ -624,6 +623,111 @@ continuous_agg_refresh_with_window(const ContinuousAgg *cagg,
...
+		/* No batching: refresh the whole window as a single batch */
+		refresh_window_list = lappend(refresh_window_list, refresh_window);
+	}
+	else
+	{
+		/* Batches are already bucket-aligned by the split function */
+		switch (context.callctx)
+		{
+			case CAGG_REFRESH_POLICY:
+				context.callctx = CAGG_REFRESH_POLICY_BATCHED;
```
Comment (gayyappan) [inline tsl/src/continuous_aggs/refresh.c:661]: why does this context matter i.e. if it is via policy or window?
we have the same behavior now, hence the question

Comment (melihmutlu) [reply tsl/src/continuous_aggs/refresh.c:661]: it's used in `emit_up_to_date_notice()` which does not print up-to-date log for policies.
